tests/sentry/backup/test_correctness.py

```diff
@@ -0,0 +1,188 @@
+import io
+import os
+from pathlib import Path
+from typing import Dict, List, NewType, Optional, Protocol, Tuple
+
+from dateutil import parser
+from django.db import IntegrityError
+from jsondiff import diff
+from pydantic import FilePath, PositiveInt
+
+from sentry.db.postgres.roles import in_test_psql_role_override
+from sentry.runner.commands.backup import exporter, importer
+from sentry.utils import json
+from sentry.utils.json import JSONData
+from sentry.utils.pytest.fixtures import django_db_all
+
+ComparatorName = NewType("ComparatorName", str)
+ModelName = NewType("ModelName", str)
+
+
+class InstanceID:
+    """Every entry in the generated backup JSON file should have a unique model+pk combination, which serves as its identifier."""
+
+    model: ModelName
+    pk: PositiveInt
+
+    def __init__(self, model: ModelName, pk: PositiveInt):
+        self.model = model
+        self.pk = pk
+
+    def __eq__(self, other):
+        return (
+            hasattr(other, "model")
+            and self.model == other.model
+            and hasattr(other, "pk")
+            and self.pk == other.pk
+        )
+
+    def __hash__(self):
+        return hash((self.model, self.pk))
+
+    def __str__(self):
+        return f'InstanceID(model: "{self.model}", pk: {self.pk})'
+
+
+class ComparatorFinding:
+    """Store all information about a single failed matching between expected and actual output."""
+
+    name: ComparatorName
+    on: InstanceID
+    reason: str
+
+    def __init__(self, name: ComparatorName, on: InstanceID, reason: Optional[str] = ""):
+        self.name = name
+        self.on = on
+        self.reason = reason
+
+    def __str__(self):
+        return (
+            f'Finding(\n\tname: "{self.name}",\n\ton: {str(self.on)},\n\treason: {self.reason}\n)'
+        )
+
+
+class ComparatorFindings(List[ComparatorFinding]):
+    """A wrapper type for a list of 'ComparatorFinding' which enables pretty-printing in asserts."""
+
+    def __str__(self):
+        out = ""
+        for finding in self:
+            out += "\n" + str(finding)
+        return out
+
+    def maybe_assert(self):
+        if len(self) > 0:
+            assert False, str(self)
+
+
+class JSONMutatingComparator(Protocol):
```

Comment:
I'm concerned this test is far too complicated -- there is a ton of logic here if you need to reach for `Protocol`s for a test.

tests should be _obvious_ what they are testing -- however this appears to be veering into the "I need tests for my tests" territory. https://testing.googleblog.com/2014/07/testing-on-toilet-dont-put-logic-in.html

Reply:
I'm not sure what we could do to make it much simpler? The overarching goal is to ensure that an import followed by immediate export of some JSON produces as idempotent a result as possible. Unfortunately, some fields will always change (update timestamps, perhaps hashes or keys, etc), so we need some more sophisticated logic to compare those fields.

We could break this workflow up into unit tests, but then we lose the property of this being a round trip integration test. Especially for something so undertested at the moment (there are zero correctness tests for backup.py!), this seems valuable to have.

I'd like to add more tests of this shape in future PRs, which will re-use this comparator code.

Reply:
you can utilize freezegun to freeze the time to make things deterministic

Reply:
Done.